R/read-gs.R

```diff
@@ -15,11 +16,200 @@ read_gs <- function(path, verbose = FALSE){
+  
+  if (source == "auto"){
+    #fill in later
+    
+    if (auto_determine == "Gradescope"){
+      attr(grades, "source") <- "Gradescope"
+      
+      return(check_data_format(grades))
+    } else if (auto_determine == "Canvas"){
+      #the read_canvas_grades does all of the lifting
+      return(read_canvas_grades(grades))
+    } else {
+      stop("Data source is unrecognized.")
+    }
+    
+  } else if (source == "Gradescope"){
+    if (source != auto_determine){
+      warning(paste0(
+        "Grade data will be processed as ", source, 
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    
+    attr(grades, "source") <- "Gradescope"
+    
+    return(check_data_format(grades))
+      
+  } else if (source == "Canvas"){
+    if (source != auto_determine){
+      warning(paste0(
+        "Grade data will be processed as ", source, 
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    #the read_canvas_grades does all of the lifting
+    return(read_canvas_grades(grades))
+    
+  } else {
+    stop(paste0(
+      "Specified source type ", source,
+      " is not currently supported."
+    ))
+  }
+  
+  
+ 
+}
+
+#'Convert Canvas grade dataframe into standard format
+#'
+#'Takes in a dataframe uploaded from Canvas. 
+#'Returns a dataframe in our standardized format (ie like Gradescope).
+#'
+#'
+#'@importFrom stringr str_extract str_match
+#'@importFrom dplyr filter select slice rename_with rename left_join
+#'@importFrom purrr discard
+#'@keywords internal
+read_canvas_grades <- function(grades){
+  #first determine which columns are assignments
+  
+  #canvas categories will not be kept
+  
+  assignments <- stringr::str_extract(names(grades),
+                                      ".+\\s*\\(\\d+\\)") |>
+                purrr::discard(is.na)
+  
+ 
+  #now reconfigure the points possible info
+  max_points <- dplyr::filter(grades, Student == "    Points Possible") |>
+    dplyr::select(all_of(assignments)) |>
+    dplyr::slice(rep.int(1, nrow(grades))) |>
+    dplyr::rename_with(function(x){
+      paste0(x, " - Max Points")
+    })
+  #add ID column to max points to facilitate matching
+  max_points$ID <- grades$ID
+  
+  #remove unneeded cols, add max_point column, and remove rows that are not students
+  grades <- grades |>
+    dplyr::select(c(all_of(assignments), "ID", "Student", 
+                    "SIS User ID", "Section")) |>
+    dplyr::left_join(max_points, 
+                     by = dplyr::join_by(ID == ID)) |>
+    dplyr::filter(!(Student %in% c("    Points Possible", "Student, Test")))
+  
+  #add columns for submission time and lateness
+  
+  sub_cols <- paste0(assignments, " - Submission Time")
+  
+  late_cols <- paste0(assignments, " - Lateness (H:M:S)")
+  
+  grades[sub_cols] <- as.POSIXct(NA)
```

Comment:
I didn't realize this: after `read_csv()` of a gradescope df, do the submission time cols read in as `POSIXct`?